aws-amplify/amplify-category-api
# feat: bootstrap for adding amplify managed table

#### Description of changes
Bootstrap for adding the new table management for DynamoDB tables capable of multiple GSIs update. 

There are still features to come like the construct wrapping the custom resource, handling non-gsi table update in lambda, etc. Those are marked as TODO and will be implemented later.

Test included in the PR:
- `getNextGSIUpdate`
-  New table resource generator (partially)

Other unit and e2e tests will be added in separate PR. 

- New table resource generator
  - Custom resource with DynamoDB table properties
  - Use Custom Resource Provider Framework
  - The params & conditions are kept in the template
  - New nested Stack `AmplifyTableManager` for the framework resources & lambda functions
- Custom resource lambda functions
  - `onEvent`
    - Handles CFN event of `Create`, `Update` and `Delete`
    - Parsed the event to the table definition input
    - Make DynamoDB SDK calls based on event type
  - `isComplete`
    - Check if the table/indexes are ready
    - Make follow-up updates if there are any
- New transformer parameter: `useAmplifyManagedTableResources`
  - ONLY for internal usage



##### CDK / CloudFormation Parameters Changed



#### Issue #, if available



#### Description of how you validated changes
The changes are validated by my local CDK app. The `AmplifyGraphqlApi` construct package is linked locally for test.

So far I've successfully tested the new table management with the following scenarios:

- Create table
- Update table with adding 2 GSIs
- Update table with deleting 2 GSIs
- Change key schema to replace the table (delete then create)
- Delete table

#### Checklist



- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Any CDK or CloudFormation parameter changes are called out explicitly

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
